internal/server/repository_workerauth.go

```diff
@@ -114,18 +112,77 @@ func (r *WorkerAuthRepositoryStorage) Store(ctx context.Context, msg nodee.Messa
 // Node information is stored in two parts:
 // * the workerAuth record is stored with a reference to a worker
 // * certificate bundles are stored with a reference to the workerAuth record and issuing root certificate
-func StoreNodeInformationTx(ctx context.Context, writer db.Writer, databaseWrapper wrapping.Wrapper, node *types.NodeInformation, _ ...Option) error {
+func StoreNodeInformationTx(ctx context.Context, reader db.Reader, writer db.Writer, kmsCache *kms.Kms, scopeId string, node *types.NodeInformation, _ ...Option) error {
 	const op = "server.(WorkerAuthRepositoryStorage).storeNodeInformation"
+	if isNil(reader) {
+		return errors.New(ctx, errors.InvalidParameter, op, "missing reader")
+	}
 	if isNil(writer) {
 		return errors.New(ctx, errors.InvalidParameter, op, "missing writer")
 	}
-	if isNil(databaseWrapper) {
-		return errors.New(ctx, errors.InvalidParameter, op, "missing database wrapper")
+	if isNil(kmsCache) {
+		return errors.New(ctx, errors.InvalidParameter, op, "missing kms")
+	}
+	if scopeId == "" {
+		return errors.New(ctx, errors.InvalidParameter, op, "missing scope id")
 	}
 	if node == nil {
 		return errors.New(ctx, errors.InvalidParameter, op, "missing NodeInformation")
 	}
 
+	databaseWrapper, err := kmsCache.GetWrapper(ctx, scopeId, kms.KeyPurposeDatabase)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+
+	var workerId string
+	if node.WrappingRegistrationFlowInfo != nil {
+		if util.IsNil(node.WrappingRegistrationFlowInfo.ApplicationSpecificParams) {
+			return errors.New(ctx, errors.InvalidParameter, op, "in wrapping registration flow but application specific parameters not provided")
+		}
+		// In this case we are using the wrapping registration flow and it's
+		// been validated; we need to upsert first to ensure that the worker
+		// exists in the database.
+		wci := new(cluster.WorkerConnectionInfo)
+		if err := mapstructure.Decode(node.WrappingRegistrationFlowInfo.ApplicationSpecificParams.AsMap(), wci); err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		if wci.Name == "" {
+			return errors.New(ctx, errors.InvalidParameter, op, "in wrapping registration flow but worker name not provided")
+		}
+		if wci.BoundaryVersion == "" {
+			return errors.New(ctx, errors.InvalidParameter, op, "in wrapping registration flow but boundary version not provided")
+		}
+		wConf := NewWorker(scopeId,
+			WithName(wci.Name),
+			WithDescription(wci.Description),
+			WithReleaseVersion(fmt.Sprintf("Boundary v%s", wci.BoundaryVersion)),
+		)
+		wConf.Type = PkiWorkerType.String()
+		wConf.PublicId, err = newWorkerIdFromScopeAndName(ctx, scopeId, wci.Name)
+		if err != nil || wConf.PublicId == "" {
+			return errors.Wrap(ctx, err, op, errors.WithMsg("error creating a worker id"))
+		}
+		workerCreateConflict := &db.OnConflict{
+			Target: db.Columns{"public_id"},
+			Action: db.SetColumns([]string{"release_version", "name", "description"}),
```

Comment:
For normal PKI workers (actually, all workers), activation state and address are sent via Status and those values can change (e.g. with a SIGHUP). I didn't see a reason to set it here since the first thing that will happen after the connection is successful is that status is sent, and it seemed prudent to keep the same paradigm across all registration methods.

Name OTOH is set via the API for normal PKI workers. I could not set release version but since that isn't subject to change (via connectivity or other changes or SIGHUP) I figured why not.